### PR TITLE
Added An Undying Pledge

### PR DESCRIPTION
**Pull request by Prathen**
**_I affirm:_**
- [x] that I agree to Topaz Next's [Limited Contributor License Agreement](https://github.com/topaz-next/topaz/blob/release/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have read the [Contributing Guide](https://github.com/topaz-next/topaz/blob/release/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/topaz-next/topaz/blob/release/CODE_OF_CONDUCT.md)
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

Made the An Undying Pledge quest available and completable.
Edited Stray Cloud to give the quest with the right fame, and play all cutscenes.
Added the NM to the proper ??? which triggers if you are on the proper step of the quest.
Cutscene and key item obtained after checking the ??? after defeating the NM.
Cutscene and quest completed on returning back to Stray Cloud, and gain the Light Buckler.

### CODE REVIEW

**Comment by wrenffxi** [comment]
Hi, and thanks for the contribution.

This PR is going to conflict with some of the work being done with SOL refactor, so I'm going to wait to do a full review and approval until that work has merged, and this PR has rebased on it.

In the meanwhile, I'll leave some general notes.

File: scripts/zones/Norg/Zone.lua [hunk 1]
```diff
@@ -7,11 +7,13 @@ local ID = require("scripts/zones/Norg/IDs")
 require("scripts/globals/conquest")
 require("scripts/globals/keyitems")
 require("scripts/globals/missions")
+require("scripts/globals/quests")
 require("scripts/globals/chocobo")
 -----------------------------------
 
 function onInitialize(zone)
     tpz.chocobo.initZone(zone)
+	zone:registerRegion(1, -24, 0, -59, -15, 1, -50)  -- Near the SSG exit
```
Comment (wrenffxi) [inline scripts/zones/Norg/Zone.lua]: Here, and in some of the other files, you've got tabs.  Project standard is 4 spaces.  There's probably a setting in your text editor to automatically replace tabs with 4 spaces.

File: scripts/zones/Norg/Zone.lua [hunk 2]
```diff
@@ -36,6 +38,15 @@ function onZoneIn(player, prevZone)
 end
 
 function onRegionEnter(player, region)
+    switch (region:GetRegionID()): caseof
+    {
+        [1] = function (x)  -- An Undying Pledge cs trigger
+            if player:getCharVar("pledgeCS") == 1 then
+			player:startEvent(226)
+			player:setCharVar("pledgeCS", 2)
```
Comment (wrenffxi) [inline scripts/zones/Norg/Zone.lua]: setting the charVar should happen in the onEventFinish function in this script, when csid == 226

File: scripts/zones/Norg/npcs/Stray_Cloud.lua [hunk 3]
```diff
@@ -1,18 +1,58 @@
...
+-----------------------------------
+
 
 function onTrade(player, npc, trade)
 end
 
 function onTrigger(player, npc)
-    player:startEvent(231)
-end
+    Pledge = player:getQuestStatus(OUTLANDS, tpz.quest.id.outlands.AN_UNDYING_PLEDGE)
```
Comment (wrenffxi) [inline scripts/zones/Norg/npcs/Stray_Cloud.lua]: recommend to name this var anUndyingPledge for project consistency, and local it.
`local anUndyingPledge = player:getQuestStatus(OUTLANDS, tpz.quest.id.outlands.AN_UNDYING_PLEDGE)`

note that this line of code will need to be changed after sol_refactor merges

File: scripts/zones/Norg/npcs/Stray_Cloud.lua [hunk 4]
```diff
@@ -1,18 +1,58 @@
...
 
 function onTrade(player, npc, trade)
 end
 
 function onTrigger(player, npc)
-    player:startEvent(231)
-end
+    Pledge = player:getQuestStatus(OUTLANDS, tpz.quest.id.outlands.AN_UNDYING_PLEDGE) 	
+
+    if (Pledge == QUEST_AVAILABLE and player:getFameLevel(NORG) >= 4) then
```
Comment (wrenffxi) [inline scripts/zones/Norg/npcs/Stray_Cloud.lua]: project standard is to not include parens unless needed, so here and in other places,
`if Pledge == QUEST_AVAILABLE and player:getFameLevel(NORG) >= 4 then`

**Comment by Era-Lusiphur** [comment]
Bringing in comments from Discord, formatting issues and utilize npc_util global.

File: scripts/zones/Norg/npcs/Stray_Cloud.lua [hunk 5]
```diff
@@ -1,18 +1,58 @@
...
+
 
 function onEventUpdate(player, csid, option)
 end
 
 function onEventFinish(player, csid, option)
+	if (csid == 225) then
+		player:addQuest(OUTLANDS, tpz.quest.id.outlands.AN_UNDYING_PLEDGE)
+		player:setCharVar("pledgeCS", 1)
+	elseif (csid == 227) then
```
Comment (wrenffxi) [inline scripts/zones/Norg/npcs/Stray_Cloud.lua]: can use npcUtil.completeQuest here, which handles inventory space checking and other stuff:

```
if csid == 225 then
    player:addQuest(OUTLANDS, tpz.quest.id.outlands.AN_UNDYING_PLEDGE)
    player:setCharVar("pledgeCS", 1)
elseif
    csid == 227 and
    npcUtil.completeQuest(player, OUTLANDS, tpz.quest.id.outlands.AN_UNDYING_PLEDGE, {
        item = 12375,
        fameArea = NORG,
        fame = 50,
        var = "pledgeCS",
    })
then
    player:delKeyItem(tpz.ki.CALIGINOUS_BLADE)
end
```

note that npcUtil.completeQuest returns doesn't do any of the completion stuff, and returns false, if player doesn't have inventory space for the item.

File: scripts/zones/Sea_Serpent_Grotto/mobs/Glyryvilu.lua [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+-- Area: Sea Serpent Grotto
+--  Mob: Glyryvilu
+-- Note: Popped by qm1
+-- !pos 135x -9y 220z
+-- Involved in Quest: An Undying Pledge
+-----------------------------------
+
+
+function onMobDeath(mob, player, isKiller)
+    player:setCharVar("pledgeNM_killed", 1)
```
Comment (wrenffxi) [inline scripts/zones/Sea_Serpent_Grotto/mobs/Glyryvilu.lua]: I'm guessing this variable should only be set on players who are currently on the quest (maybe even on a particular stage of the quest).

File: scripts/zones/Sea_Serpent_Grotto/Zone.lua [hunk 7]
```diff
@@ -36,4 +38,10 @@ function onEventUpdate(player, csid, option)
 end
 
 function onEventFinish(player, csid, option)
+	if (csid == 18) then
```
Comment (wrenffxi) [inline scripts/zones/Sea_Serpent_Grotto/Zone.lua]: this event finish should happen in the qm5.lua script that starts event 18

File: scripts/zones/Sea_Serpent_Grotto/npcs/qm5.lua [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+require("scripts/globals/keyitems")
+require("scripts/globals/settings")
+require("scripts/globals/quests")
+require("scripts/globals/npc_util")
+-----------------------------------
+
+function onTrigger(player, npc)
+	if player:getCharVar("pledgeCS") == 2 and player:getCharVar("pledgeNM_killed") == 1 then
+		player:startEvent(18)
+	elseif player:getCharVar("pledgeCS") == 2 and player:getCharVar("pledgeNM_killed") == 0 then
```
Comment (wrenffxi) [inline scripts/zones/Sea_Serpent_Grotto/npcs/qm5.lua]:
```
if player:getCharVar("pledgeCS") == 2 and player:getCharVar("pledgeNM_killed") == 1 then
    player:startEvent(18)
elseif
    player:getCharVar("pledgeCS") == 2 and
    player:getCharVar("pledgeNM_killed") == 0 and
    npcUtil.popFromQM(player, npc, ID.mob.GLYRYVILU, {hide = 0})    
then
    player:messageSpecial(ID.text.BODY_NUMB_DREAD)
else
    player:messageSpecial(ID.text.NOTHING_OUT_OF_ORDINARY)
end
```

note that popFromQM returns false if the mob is already spawned
also, the claim param defaults to true, so it's not needed

File: scripts/zones/Norg/Zone.lua [hunk 9]
```diff
@@ -7,11 +7,13 @@ local ID = require("scripts/zones/Norg/IDs")
 require("scripts/globals/conquest")
 require("scripts/globals/keyitems")
 require("scripts/globals/missions")
+require("scripts/globals/quests")
```
Comment (wrenffxi) [inline scripts/zones/Norg/Zone.lua]: unused require; not needed.

note that charVars can be accessed without quests require.  quests require is needed for references such as tpz.quest.id.foo.FOO or QUEST_AVAILABLE.

File: scripts/zones/Norg/Zone.lua [hunk 10]
```diff
@@ -36,6 +38,14 @@ function onZoneIn(player, prevZone)
 end
 
 function onRegionEnter(player, region)
+    switch (region:GetRegionID()): caseof
+    {
+        [1] = function (x)  -- An Undying Pledge cs trigger
+            if player:getCharVar("anUndyingPledgeCS") == 1 then
+            player:startEvent(226)
```
Comment (wrenffxi) [inline scripts/zones/Norg/Zone.lua]: indentation needed.  i notice some of these files still have tabs in them.  look in your text editor for a setting that's something like "convert existing tabs into spaces", then open and save each file you touched with this PR.  that should convert all the tabs.

note that "convert existing tabs into spaces" is not the same thing as "convert new tabs into spaces" -- editor will probably have two different settings.  I'd turn both on for this project.  Also useful is the setting to "trim trailing whitespace at end of lines".

File: scripts/zones/Norg/npcs/Stray_Cloud.lua [hunk 11]
```diff
@@ -1,18 +1,58 @@
...
-    player:startEvent(231)
-end
+    local anUndyingPledge = player:getQuestStatus(OUTLANDS, tpz.quest.id.outlands.AN_UNDYING_PLEDGE) 	
+
+    if anUndyingPledge == QUEST_AVAILABLE and player:getFameLevel(NORG) >= 4 then
+        player:startEvent(225) -- Start quest
+    elseif anUndyingPledge == QUEST_ACCEPTED and player:hasKeyItem(tpz.ki.CALIGINOUS_BLADE) then
+        player:startEvent(227) -- Quest Finish
+    elseif anUndyingPledge == QUEST_ACCEPTED and player:getCharVar("anUndyingPledgeCS") == 1 then	
+	    player:startEvent(228) -- Extra Dialogue
```
Comment (wrenffxi) [inline scripts/zones/Norg/npcs/Stray_Cloud.lua]: indentation

File: scripts/zones/Sea_Serpent_Grotto/IDs.lua [hunk 12]
```diff
@@ -37,6 +37,7 @@ zones[tpz.zone.SEA_SERPENT_GROTTO] =
         GOLD_CHECK                 = 7347,  -- You see something gold glittering around the indentation.
         COMPLETED_CHECK            = 7348,  -- It is a door you can open using <item>!
         SENSE_OMINOUS_PRESENCE     = 7362,  -- You sense an ominous presence...
+        BODY_NUMB_DREAD			   = 7550,  -- Your body goes numb with dread!
```
Comment (wrenffxi) [inline scripts/zones/Sea_Serpent_Grotto/IDs.lua]: tabs -> spaces

File: scripts/zones/Sea_Serpent_Grotto/IDs.lua [hunk 13]
```diff
@@ -118,6 +119,7 @@ zones[tpz.zone.SEA_SERPENT_GROTTO] =
         },
         MIMIC        = 17498564,
         WATER_LEAPER = 17498565,
+        GLYRYVILU 	 = 17498566,
```
Comment (wrenffxi) [inline scripts/zones/Sea_Serpent_Grotto/IDs.lua]: tabs -> spaces

File: scripts/zones/Sea_Serpent_Grotto/Zone.lua [hunk 14]
```diff
@@ -6,6 +6,8 @@
 local ID = require("scripts/zones/Sea_Serpent_Grotto/IDs")
 require("scripts/globals/conquest")
 require("scripts/globals/treasure")
+require("scripts/globals/zone")
```
Comment (wrenffxi) [inline scripts/zones/Sea_Serpent_Grotto/Zone.lua]: both of these requires are unused; not needed

File: scripts/zones/Sea_Serpent_Grotto/mobs/Glyryvilu.lua [hunk 15]
```diff
@@ -0,0 +1,14 @@
+-----------------------------------
+-- Area: Sea Serpent Grotto
+--  Mob: Glyryvilu
+-- Note: Popped by qm1
+-- !pos 135x -9y 220z
```
Comment (wrenffxi) [inline scripts/zones/Sea_Serpent_Grotto/mobs/Glyryvilu.lua]: `!pos 135 -9 220 176`

this way someone can copy/paste the command into their game to warp there.  note 176 is the zone id

File: scripts/zones/Sea_Serpent_Grotto/npcs/qm5.lua [hunk 16]
```diff
@@ -0,0 +1,36 @@
+-----------------------------------
+-- Area: Sea Serpent Grotto
+--  NPC: qm1 (???)
+-- Quests: An Undying Pledge
+-- !pos 135x -9y 220z
```
Comment (wrenffxi) [inline scripts/zones/Sea_Serpent_Grotto/npcs/qm5.lua]: `!pos 135 -9 220 176`

File: scripts/zones/Sea_Serpent_Grotto/npcs/qm5.lua [hunk 17]
```diff
@@ -0,0 +1,36 @@
+-----------------------------------
+-- Area: Sea Serpent Grotto
+--  NPC: qm1 (???)
```
Comment (wrenffxi) [inline scripts/zones/Sea_Serpent_Grotto/npcs/qm5.lua]: qm5

File: scripts/zones/Sea_Serpent_Grotto/npcs/qm5.lua [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+
+
+function onEventFinish(player, csid, option)
+    if (csid == 18) then
+        player:addKeyItem(tpz.ki.CALIGINOUS_BLADE)
+        player:messageSpecial(ID.text.KEYITEM_OBTAINED, tpz.ki.CALIGINOUS_BLADE)
+        player:setCharVar("anUndyingPledgeCS", 3)
+        player:setCharVar("anUndyingPledgeNM_killed", 0)
+    end
+end
```
Comment (wrenffxi) [inline scripts/zones/Sea_Serpent_Grotto/npcs/qm5.lua]: project standard: add newline at end of file

File: scripts/zones/Sea_Serpent_Grotto/npcs/qm5.lua [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+    then
+        player:messageSpecial(ID.text.BODY_NUMB_DREAD)
+    else
+        player:messageSpecial(ID.text.NOTHING_OUT_OF_ORDINARY)
+    end
+end
+
+
+function onEventFinish(player, csid, option)
+    if (csid == 18) then
```
Comment (wrenffxi) [inline scripts/zones/Sea_Serpent_Grotto/npcs/qm5.lua]: unneeded parens.
`if csid == 18 then`

File: scripts/zones/Sea_Serpent_Grotto/npcs/qm5.lua [hunk 20]
```diff
@@ -0,0 +1,36 @@
+-----------------------------------
+-- Area: Sea Serpent Grotto
+--  NPC: qm1 (???)
+-- Quests: An Undying Pledge
+-- !pos 135x -9y 220z
+-----------------------------------
+local ID = require("scripts/zones/Sea_Serpent_Grotto/IDs")
+require("scripts/globals/keyitems")
+require("scripts/globals/settings")
```
Comment (wrenffxi) [inline scripts/zones/Sea_Serpent_Grotto/npcs/qm5.lua]: settings and quests requires are unused; not needed

File: scripts/zones/Sea_Serpent_Grotto/mobs/Glyryvilu.lua [hunk 21]
```diff
@@ -0,0 +1,14 @@
+-----------------------------------
+-- Area: Sea Serpent Grotto
+--  Mob: Glyryvilu
+-- Note: Popped by qm1
```
Comment (wrenffxi) [inline scripts/zones/Sea_Serpent_Grotto/mobs/Glyryvilu.lua]: popped by qm5

File: scripts/zones/Norg/npcs/Stray_Cloud.lua [hunk 22]
```diff
@@ -1,18 +1,58 @@
 -----------------------------------
 -- Area: Norg
 --  NPC: Stray Cloud
--- Standard Info NPC
+-- Starts and Ends Quest: An Undying Pledge
+-- !pos-20.617, 1.097, -29.165, 133
 -----------------------------------
+require("scripts/globals/titles")
```
Comment (wrenffxi) [inline scripts/zones/Norg/npcs/Stray_Cloud.lua]: titles, settings, and NorgIDs requires are not used in this script, so aren't needed

**Comment by zach2good** [comment]
Hello and welcome!

Yesterday we just started a refactoring round where we're going to modify every mob script so we won't be able to merge this until that work is complete. Luckily, it's going well and it _should_ be complete in less than a month.

So I'd say continue working through the review issues Wren has raised for you, and make sure you're familiar with the [Style Guide](https://github.com/topaz-next/topaz/blob/release/CONTRIBUTING.md#style-guide); specifically the "all-languages" and "Lua" sections. I'm going to slap a "hold" sticker on this PR, don't be upset that it's big and red, it just means that it shouldn't be merged until other conditions are met.

👍

**Review by zach2good — APPROVED** [review]
Thanks for taking on all the feedback from Wren, ready to merge 👍 :)

**Pull request merged**